src/assembler.jl

```diff
@@ -89,71 +100,94 @@ Assembles the element residual `ge` into the global residual vector `g`.
+
+"""
+Assembler for symmetric sparse matrix with CSC storage type.
+"""
+struct SymmetricCSCAssembler{Tv,Ti, MT <: Symmetric{Tv,<:AbstractSparseMatrixCSC{Tv,Ti}}} <: AbstractSparseCSCAssembler
+    K::MT
     f::Vector{Tv}
     permutation::Vector{Int}
     sorteddofs::Vector{Int}
 end
 
-function Base.show(io::IO, ::MIME"text/plain", a::Union{AssemblerSparsityPattern,AssemblerSymmetricSparsityPattern})
+function Base.show(io::IO, ::MIME"text/plain", a::AbstractSparseAssembler)
     print(io, typeof(a), " for assembling into:\n - ")
-    summary(io, a.K)
-    if !isempty(a.f)
+    summary(io, get_matrix(a))
+    f = get_vector(a)
+    if !isempty(f)
         print(io, "\n - ")
-        summary(io, a.f)
+        summary(io, f)
     end
 end
 
-matrix_handle(a::AssemblerSparsityPattern) = a.K
-matrix_handle(a::AssemblerSymmetricSparsityPattern) = a.K.data
-vector_handle(a::Union{AssemblerSparsityPattern, AssemblerSymmetricSparsityPattern}) = a.f
+matrix_handle(a::AbstractSparseCSCAssembler) = a.K
+matrix_handle(a::SymmetricCSCAssembler) = a.K.data
+vector_handle(a::AbstractSparseCSCAssembler) = a.f
 
 """
-    start_assemble(K::SparseMatrixCSC;            fillzero::Bool=true) -> AssemblerSparsityPattern
-    start_assemble(K::SparseMatrixCSC, f::Vector; fillzero::Bool=true) -> AssemblerSparsityPattern
+    start_assemble(K::AbstractSparseMatrixCSC;            fillzero::Bool=true, maxcelldofs_hint::Int=0) -> CSCAssembler
+    start_assemble(K::AbstractSparseMatrixCSC, f::Vector; fillzero::Bool=true, maxcelldofs_hint::Int=0) -> CSCAssembler
 
-Create a `AssemblerSparsityPattern` from the matrix `K` and optional vector `f`.
+Create a `CSCAssembler` from the matrix `K` and optional vector `f`.
 
-    start_assemble(K::Symmetric{SparseMatrixCSC};                 fillzero::Bool=true) -> AssemblerSymmetricSparsityPattern
-    start_assemble(K::Symmetric{SparseMatrixCSC}, f::Vector=Td[]; fillzero::Bool=true) -> AssemblerSymmetricSparsityPattern
+    start_assemble(K::Symmetric{AbstractSparseMatrixCSC};                 fillzero::Bool=true, maxcelldofs_hint::Int=0) -> SymmetricCSCAssembler
+    start_assemble(K::Symmetric{AbstractSparseMatrixCSC}, f::Vector=Td[]; fillzero::Bool=true, maxcelldofs_hint::Int=0) -> SymmetricCSCAssembler
 
-Create a `AssemblerSymmetricSparsityPattern` from the matrix `K` and optional vector `f`.
+Create a `SymmetricCSCAssembler` from the matrix `K` and optional vector `f`.
 
-`AssemblerSparsityPattern` and `AssemblerSymmetricSparsityPattern` allocate workspace
+`CSCAssembler` and `SymmetricCSCAssembler` allocate workspace
 necessary for efficient matrix assembly. To assemble the contribution from an element, use
 [`assemble!`](@ref).
 
 The keyword argument `fillzero` can be set to `false` if `K` and `f` should not be zeroed
 out, but instead keep their current values.
+The keyword argument `maxcelldofs_hint` can be set to the largest number celldofs over all cells
+to eliminate runtime allocations.
```

Comment:
I guess technically they could still happen due to the resize, just unlikely? Perhaps "reduce" or "reduce the chance of"

Reply:
If you set it to the described value you should have a hard guarantee that resize is basically a noop.

Reply:
Seems hard to trigger at least, somehow filling the values seems to be required to be even more safe, but I don't know the internals good enough to say what is happening. 
```julia
vals = Float64[];
resize!(vals, 100_000_000);
@time resize!(vals, 1);
# 0.000011 seconds
GC.gc();
@time resize!(vals, 100_000_000);
# 0.000010 seconds
@time fill!(vals, 1.0);
# 0.491245 seconds (133 allocations: 6.391 KiB, 2.23% compilation time)
@time resize!(vals, 100_000_000_0);
# 1.977564 seconds (1 allocation: 0 bytes)
```
where the last one is just interesting, perhaps a bug in reporting...

Reply:
cc @fredrikekre can you bring light into the dark here? From my understanding of the `resize!` docs we have a guarantee that once we have an vector with $n$ elements consecutive resize calls will not allocate as long as the resize is to at most $n$ elements.

Reply:
I think currently there is no shrinking of the backing memory if that's what you are asking. Is the potential runtime resize problematic though? Perhaps this kwarg isn't necesary.

Reply:
This is prospective for GPU assembly where we are not allowed to runtime allocate memory.